syntax = "proto3";

package dedup;

// ---------- Search (online) ----------

message SearchRequest {
  int64 query_id = 1;
  bytes embedding_f32 = 2;
}

message SearchBatchRequest {
  repeated int64 query_ids = 1;
  repeated bytes embeddings_f32 = 2;
}

message SearchResponse {
  int64 query_id = 1;
  int64 nearest_id = 2;
  float distance_sq = 3;
  float distance_l2 = 4;
  bool is_match = 5;

  double faiss_ms = 6;
  double faiss_batch_ms = 7;
  int32 batch_size = 8;

  double aggregator_wait_ms = 9;
  string note = 10;
}

message SearchBatchResponse {
  repeated SearchResponse results = 1;
  double faiss_batch_ms = 2;
  int32 batch_size = 3;
}

// ---------- Ingest (index updates) ----------

message IngestBatchRequest {
  repeated int64 ids = 1;
  repeated bytes embeddings_f32 = 2;
}

message IngestBatchResponse {
  int32 accepted = 1;
  int32 queued = 2;
  double enqueue_ms = 3;
  string note = 4;
}

service FaissDedup {
  rpc Search(SearchRequest) returns (SearchResponse);
  rpc SearchBatch(SearchBatchRequest) returns (SearchBatchResponse);
  rpc IngestBatch(IngestBatchRequest) returns (IngestBatchResponse);
}

service FaissAggregator {
  rpc Search(SearchRequest) returns (SearchResponse);
}
